### PR TITLE
[ISSUE-143] students have many active volunteers

### LINKED ISSUE #143: Students have many active volunteers
# Summary

After meeting with the stakeholders again, we've discovered that some students have many active volunteers at one time. The student_volunteer_assignment logic needs to be updated to account for this

## Things to Consider

This product is still pre-MVP. 

# Criteria for Completion
 - [ ] When adding/updating a student, multiple active volunteers can be assigned
 - [ ] When the form is saved, current assigned volunteers stay assigned UNLESS they've been removed.
 - [ ] When the form is save, additional volunteers will be assigned to the student AND active. (validation to only have one assigned volunteer must change)

What student form looks like now:

Screen Shot 2021-11-20 at 10 20 26 AM

What we eventually would like it to work/look like:

Screen Shot 2021-11-20 at 10 18 17 AM

This story does not need to accomplish everything. But the ultimate goal is to have a typeahead input. And when the end-user clicks on a volunteer name, it is added to the list below. In the list below, a volunteer can be removed by clicking the X next to there name.

This story is acceptable when any functionality exists on this form that allows for multiple volunteer assignment.

Questions? - find us on Slack :).

### PR DESCRIPTION
**Pull request by goromlagche**
Resolves #143  

### Description

- Include summary of change and what issue is fixed, including relevant motivation and context.

I have added changes to allow assigning multiple volunteers to students.
I also have added a new js library to search and select multiple volunteers on the UI

- Explain the need for any added dependencies for this change. (gems, js libraries, etc.)

We need a new library for having auto-complete on searching for volunteers.
I have used [slim-select](https://github.com/excid3/stimulus-slimselect) for this.

- How have you tested this?

Yes, please find the attached gif. I also have added specs to handle assigning multiple volunteers to a student.

- Anything else we should know about?

Please review the PR thoroughly, since I am new to this codebase :)

### Type of change

* New feature (non-breaking change which adds functionality)

### Screenshots

student-mutiple-volunteer

### CODE REVIEW

File: app/models/student.rb [hunk 1]
```diff
@@ -45,13 +45,25 @@ def activate!
...
+
+    new_volunteers = volunteer_ids - active_volunteer_ids
+    if new_volunteers.present?
+      StudentVolunteerAssignment
+        .insert_all!(new_volunteers.map { |volunteer_id|
+                       {student_id: id,
+                        volunteer_id: volunteer_id,
+                        created_at: Time.zone.now,
+                        updated_at: Time.zone.now}
+                     })
```
Comment (garettarrowood) [inline app/models/student.rb:66]: Interesting! I wasn't familiar with `insert_all`. Looks like it avoids callbacks and manually requires setting timestamps. Let's not use this here.

At CEP, most students only have one volunteer. But there are a few with 2 or 3. So I don't think we'd hit a performance issue that requires us to need this speedier update, and sacrifice validations/callbacks.

File: app/models/student_volunteer_assignment.rb [hunk 2]
```diff
@@ -1,8 +1,9 @@
 class StudentVolunteerAssignment < ApplicationRecord
+  attribute :start_date, default: -> { Date.current }
```
Comment (garettarrowood) [inline app/models/student_volunteer_assignment.rb:2]: Nice!

**Comment by garettarrowood** [comment]
Great work here! Thank you for this contribution. Particularly adding a lightweight typeahead library.

Change the update strategy in the one spot I noted and this is good to merge.

File: app/models/student.rb (hunk 1 above)
Comment (goromlagche) [reply app/models/student.rb]: ack. Thank you for sharing the info. I have done the update. Please check.

**Review by garettarrowood — APPROVED** [review]

**Pull request merged**